### PR TITLE
3x3 winograd convs

### PR DESCRIPTION
**Pull request by chaosagent**
New pr, old: #1037 . Tracking issue: #1618 . This is the minimal set of changes for fast winograd. Compilation time is slow with only this patch, that needs #1484 and some expr_idxs changes.

On rtx3090:

```
conv bs:  4 chans:128 -> 128 k:3              0.70 ms (86888.78 GFLOPS 126888.59 GB/s) in torch,    0.79 ms (77287.86 GFLOPS 112867.82 GB/s) in tinygrad,    1.12x slower   60806.07 MOPS 88798.54 MB
```

It can be 0.77ms with a slightly different hand coded optimization change, or it can be very good with KOPT.

We can merge once this part of the code is refined and the compile time is sufficiently low.

### CODE REVIEW

**Comment by tinyb0t** [comment]
Changes made in `tinygrad/`:
```
------------------------------------------------------------
files                             insertions       deletions
------------------------------------------------------------
tinygrad/codegen/optimizer.py             30               9
tinygrad/mlops.py                          4               0
tinygrad/shape/shapetracker.py             4               0
tinygrad/tensor.py                        32               4
------------------------------------------------------------
total                                     70              13
------------------------------------------------------------
lines added in the tinygrad folder: 57
```

File: tinygrad/tensor.py [hunk 1]
```diff
@@ -503,11 +503,38 @@ def conv2d(self, weight:Tensor, bias:Optional[Tensor]=None, groups=1, stride=1,
...
+      # compute 6x6 winograd tiles: GgGt, BtdB
+      gfactors = apply_matrix(winograd_G, g).contiguous()  # (HWI, bs=1, groups, rcout, cin, tyx=(1,1))
+      dfactors = apply_matrix(winograd_Bt, d).contiguous().reshape(*HWI, bs, groups, 1, cin, *tyx)  # (HWI, bs, cin_, tyx) -> (HWI, bs, groups, 1 ,cin, *tyx)
+
+      ret = apply_matrix(winograd_At, (gfactors * dfactors).sum(axis=-1-len(HW)))  # matmul; sum across cin: (HWI, bs, groups, rcout, *tyx); then HWI -> HWO: (HWO, bs, groups, rcout, *tyx)
+
+      ret = ret.permute([*range(len(HW), len(ret.shape)-len(HW)), *[i+o for i in range(len(HW)) for o in [len(ret.shape)-len(HW),0]]])  # interleave tyx and HWO: (bs, groups, rcout, oy, HO, ox, WO)
+      ret = ret.reshape(bs, cout, *[c * HWO[i] for i, c in enumerate(tyx)]).shrink(tuple((0, s) for s in [bs, cout, *oyx]))  # merge groups and rcout, tyx and HWO: (bs, groups, cout, *yx), shrink to final
+
+      return (ret if bias is None else ret.add(bias.reshape(1, -1, *[1 for _ in range(len(HW))]))).contiguous().contiguous_backward()
```
Comment (geohot) [inline tinygrad/tensor.py:537]: Why is contiguous needed here?

Comment (chaosagent) [reply tinygrad/tensor.py:537]: This kernel has 36 inputs (72 with bias), it cannot take many more kernel fusions. Ideally we would infer in lazy.py.

**Comment by geohot** [comment]
I see, tests are failing because it's slow.

I should be able to finish up the UAst stuff by the end of the week, that'll be a much better starting point for making the linearizer fast.

**Comment by geohot** [comment]
Want to merge this with:
* winograd off by default
* enabled with WINO=1
* a few small tests to ensure correctness, okay if they take 30 seconds to compile a kernel

This way I can start playing with making the Linearizer fast in the tokenless branch

**Comment by chaosagent** [comment]
OK, two rudimentary tests are in. Maybe another test for no exploding kernel fusions might be nice though.

**Comment by geohot** [comment]
Hmm, how are other things failing besides torch? The tests look gated

**Comment by geohot** [comment]
I understand the compile time is slow, but this is still slow in execution on M1:

With WINO=1 (something clearly broken in tracker):
`conv bs:  4 chans:128 -> 128 k:3              2.08 ms (29192.17 GFLOPS 42630.98 GB/s) in torch,    3.51 ms (17302.10 GFLOPS 25267.23 GB/s) in tinygrad,    1.69x slower   60806.07 MOPS 88798.54 MB`

Without:
`conv bs:  4 chans:128 -> 128 k:3              2.25 ms ( 8597.74 GFLOPS    30.59 GB/s) in torch,    4.22 ms ( 4583.83 GFLOPS    16.31 GB/s) in tinygrad,    1.88x slower   19327.35 MOPS    68.76 MB`

**Comment by geohot** [comment]
Hmm, I ran again with `DEBUG=4 WINO=1 MPS=1 BIG=1 python3 test/test_speed_v_torch.py TestBigSpeed.test_large_conv_3x3` and just this one: `conv bs:  4 chans:128 -> 128 k:3              2.97 ms (20441.30 GFLOPS 29851.59 GB/s) in torch,    2.46 ms (24686.55 GFLOPS 36051.17 GB/s) in tinygrad,    0.83x faster   60806.07 MOPS 88798.54 MB` that's a lot better.

**Pull request merged**